src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -163,24 +165,27 @@ static void ReorderTaskPlacementsByTaskAssignmentPolicy(Job *job,
 
 /*
  * CreateRouterPlan attempts to create a router executor plan for the given
- * SELECT statement.  If planning fails either NULL is returned, or
- * ->planningError is set to a description of the failure.
+ * SELECT statement.  ->planningError is set if planning fails.
  */
 DistributedPlan *
 CreateRouterPlan(Query *originalQuery, Query *query,
 				 PlannerRestrictionContext *plannerRestrictionContext)
 {
-	if (MultiRouterPlannableQuery(query))
+	DistributedPlan *distributedPlan = CitusMakeNode(DistributedPlan);
+
+	MultiRouterPlannableQuery(distributedPlan, query);
+
+	if (distributedPlan->planningError == NULL)
 	{
-		return CreateSingleTaskRouterPlan(originalQuery, query,
-										  plannerRestrictionContext);
+		distributedPlan->planningError = ErrorIfQueryHasModifyingCTE(query);
```

Comment:
I know the TODO suggested moving it here, but what about putting it in `MultiRouterPlannableQuery` now that it neatly returns an error?